### PR TITLE
Build aggregate schema in Aggregate::try_new

### PR DESCRIPTION
**Pull request by andygrove**
# Which issue does this PR close?

Closes https://github.com/apache/arrow-datafusion/issues/3738

 # Rationale for this change

We had duplicate code for building an aggregate schema. This code is now inside `Aggregate::try_new` so we don't have to duplicate it everwhere,

# What changes are included in this PR?

Refactoring to eliminate duplicate code for building aggregate schema

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/expr/src/logical_plan/builder.rs [hunk 1]
```diff
@@ -700,20 +698,10 @@ impl LogicalPlanBuilder {
...
         let aggr_expr = normalize_cols(aggr_expr, &self.plan)?;
-
-        let grouping_expr: Vec<Expr> = grouping_set_to_exprlist(group_expr.as_slice())?;
-
-        let all_expr = grouping_expr.iter().chain(aggr_expr.iter());
-        validate_unique_names("Aggregations", all_expr.clone())?;
-        let aggr_schema = DFSchema::new_with_metadata(
-            exprlist_to_fields(all_expr, &self.plan)?,
-            self.plan.schema().metadata().clone(),
-        )?;
```
Comment (andygrove) [inline datafusion/expr/src/logical_plan/builder.rs:711]: This code is moved into `Aggregate::try_new`

File: datafusion/expr/src/logical_plan/plan.rs [hunk 2]
```diff
@@ -1338,6 +1341,21 @@ impl Aggregate {
...
         group_expr: Vec<Expr>,
         aggr_expr: Vec<Expr>,
+    ) -> datafusion_common::Result<Self> {
+        let grouping_expr: Vec<Expr> = grouping_set_to_exprlist(group_expr.as_slice())?;
+        let all_expr = grouping_expr.iter().chain(aggr_expr.iter());
+        validate_unique_names("Aggregations", all_expr.clone())?;
+        let schema = DFSchema::new_with_metadata(
+            exprlist_to_fields(all_expr, &input)?,
+            input.schema().metadata().clone(),
+        )?;
```
Comment (andygrove) [inline datafusion/expr/src/logical_plan/plan.rs:1351]: This is the code moved from `LogicalPlanBuilder::aggregate`

File: datafusion/optimizer/src/projection_push_down.rs [hunk 3]
```diff
@@ -332,16 +330,6 @@ fn optimize_plan(
...
 
-            let new_schema = DFSchema::new_with_metadata(
-                schema
-                    .fields()
-                    .iter()
-                    .filter(|x| new_required_columns.contains(&x.qualified_column()))
-                    .cloned()
-                    .collect(),
-                schema.metadata().clone(),
-            )?;
```
Comment (andygrove) [inline datafusion/optimizer/src/projection_push_down.rs:343]: This code was problematic and is now replaced with the code used elsewhere

Comment (alamb) [reply datafusion/optimizer/src/projection_push_down.rs:343]: yes I agree having the caller have to specify (correctly) the aggregate schema is a recipe for disaster

File: datafusion/expr/src/logical_plan/plan.rs [hunk 4]
```diff
@@ -1334,10 +1337,28 @@ pub struct Aggregate {
...
+        Self::try_new_with_schema(input, group_expr, aggr_expr, Arc::new(schema))
+    }
+
+    /// Create a new aggregate operator using the provided schema to avoid the overhead of
+    /// building the schema again when the schema is already known.
+    pub fn try_new_with_schema(
```
Comment (alamb) [inline datafusion/expr/src/logical_plan/plan.rs:1358]: I recommend making this code non pub as to use it requires a lot of knowledge about how to use it.

If we have to make it pub to call from other crates, I recommend putting a big warning on it telling others not to use it as it is intended for internal use

```suggestion
   fn try_new_with_schema(
```

**Review by alamb — APPROVED** [review]
love it

**Comment by alamb** [comment]
CI failure appears unrelated: https://github.com/apache/arrow-datafusion/issues/3743

**Pull request merged**